src/nginx/grpc_server_call.cc

```diff
@@ -691,6 +694,37 @@ bool NgxEspGrpcServerCall::TryReadDownstreamMessage() {
   return true;
 }
 
+// Allocate the chain link and buffer.
+ngx_chain_t *NgxEspGrpcServerCall::AllocNgxBufChain(size_t buflen) {
+  ngx_chain_t *cl = ngx_chain_get_free_buf(r_->pool, &buf_free_);
+  if (!cl) {
+    return nullptr;
+  }
+  ngx_buf_t *buf = cl->buf;
+
+  if (buflen > size_t(buf->end - buf->start)) {
+    // Abandon the small buf, allocate a bigger one.
+    buf = ngx_create_temp_buf(r_->pool, buflen);
+    if (!buf) {
+      return nullptr;
+    }
+    cl->buf = buf;
+  }
+
+  // Clear the data except start, end
+  u_char *start = buf->start;
+  u_char *end = buf->end;
+  ngx_memzero(buf, sizeof(ngx_buf_t));
+  buf->start = start;
+  buf->end = end;
+
+  buf->pos = buf->start;
+  buf->last = buf->start;
+  buf->temporary = 1;
```

Comment:
Does the buffer still counts as temporary if we are getting it from `ngx_chain_get_free_buf` (i.e. buflen was large enugh and we didn't call `ngx_create_temp_buf`)?

Reply:
Yes,  this temporary is used by other Nginx code to indicate the data is mutable or not per my understanding,  not referring to the buffer.